// sqiggl/dist/src/actions.js
var parser_1 = require('./parser');
/**
 * @internal
 */
exports.If = {
    key: 'if',
    rule: function (expressionResult, variables, scope, parser) { return expressionResult.value ? parser.parse(scope, variables) : null; }
};
/**
 * @internal
 */
exports.EndIf = {
    key: 'endif',
    dependents: [exports.If],
};
/**
 * @internal
 */
exports.Unless = {
    key: 'unless',
    rule: function (expressionResult, variables, scope, parser) { return !expressionResult.value ? parser.parse(scope, variables) : null; }
};
/**
 * @internal
 */
exports.EndUnless = {
    key: 'endunless',
    dependents: [exports.Unless],
};
/**
 * @internal
 */
exports.Else = {
    key: 'else',
    dependents: [exports.If, exports.Unless],
    rule: function (expressionResult, variables, scope, parser) { return parser.parse(scope, variables); }
};
/**
 * @internal
 */
exports.For = {
    key: 'for',
    rule: function (expressionResult, variables, scope, parser) {
        if (variables === void 0) { variables = {}; }
        var output = [];
        for (var _i = 0, _a = expressionResult.value; _i < _a.length; _i++) {
            var value = _a[_i];
            variables[expressionResult.iterable.local] = value;
            output.push(parser.parse(scope, variables));
        }
        return output.join(parser_1.Parser.resolveValue(expressionResult.iterable.joiner, variables) + " ");
    }
};
/**
 * @internal
 */
exports.EndFor = {
    key: 'endfor',
    dependents: [exports.For]
};
/**
 * @internal
 */
exports.End = {
    key: 'end',
    dependents: [exports.If, exports.Unless, exports.For]
};
exports.CORE_ACTIONS = [
    exports.If,
    exports.Else,
    exports.EndIf,
    exports.Unless,
    exports.EndUnless,
    exports.For,
    exports.EndFor,
    exports.End
];
